src/services/formatting/rules.ts

```diff
@@ -649,6 +662,28 @@ module ts.formatting {
             return context.TokensAreOnSameLine();
         }
 
+        static IsEndOfDecoratorContextOnSameLine(context: FormattingContext): boolean {
+            return context.TokensAreOnSameLine() &&
+                context.contextNode.decorators &&
+                Rules.NodeIsInDecoratorContext(context.currentTokenParent) &&
+                !Rules.NodeIsInDecoratorContext(context.nextTokenParent);
+        }
+
+        static NodeIsInDecoratorContext(node: Node): boolean {
+            if (node.parserContextFlags & ParserContextFlags.Decorator) {
+                return true;
+            }
+            while (node) {
+                if (isExpression(node)) {

```

Comment:
If we ever add support for decorators on class expressions, function expressions, or arrow functions then hasAncestor/getAncestor is unsafe to use. Consider the following:

```TypeScript
@class { /*A*/ } class C { }
```
At *A* above, a call to `getAncestor(SyntaxKind.Decorator)` would return the decorator for `C`, but it should have stopped at the class body. `hasAncestor` and `getAncestor` don't have a boundary. We could add a boundary check, but this is easy enough.

I may change this to the following, though:
```TypeScript
        static NodeIsInDecoratorContext(node: Node): boolean {
            while (isExpression(node)) {
                node = node.parent;
            }
            return node.kind === SyntaxKind.Decorator;
        }
```